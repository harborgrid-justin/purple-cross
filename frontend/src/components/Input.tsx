import { InputHTMLAttributes, forwardRef } from 'react';
import clsx from 'clsx';

interface InputProps extends InputHTMLAttributes<HTMLInputElement> {
  label?: string;
  error?: string;
  helperText?: string;
  fullWidth?: boolean;
}

const Input = forwardRef<HTMLInputElement, InputProps>(
  ({ label, error, helperText, fullWidth = false, className, id, ...props }, ref) => {
    const inputId = id || `input-${Math.random().toString(36).substring(7)}`;

    return (
      <div className={clsx('form-group', { 'form-group-full-width': fullWidth })}>
        {label && (
          <label htmlFor={inputId} className="form-label">
            {label}
            {props.required && <span className="form-required">*</span>}
          </label>
        )}
        <input
          ref={ref}
          id={inputId}
          className={clsx('form-input', { 'form-input-error': error }, className)}
          aria-invalid={error ? 'true' : 'false'}
          aria-describedby={
            error ? `${inputId}-error` : helperText ? `${inputId}-helper` : undefined
          }
          {...props}
        />
        {error && (
          <p id={`${inputId}-error`} className="form-error" role="alert">
            {error}
          </p>
        )}
        {helperText && !error && (
          <p id={`${inputId}-helper`} className="form-helper">
            {helperText}
          </p>
        )}
      </div>
    );
  }
);

Input.displayName = 'Input';

export default Input;
