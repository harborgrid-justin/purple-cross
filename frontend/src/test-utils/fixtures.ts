/**
 * WF-COMP-XXX | fixtures.ts - Test Fixtures
 * Purpose: Mock data fixtures for testing
 * Dependencies: None
 * Last Updated: 2025-10-24 | File Type: .ts
 */

/**
 * Mock patient data
 */
export const mockPatient = {
  id: '1',
  name: 'Buddy',
  species: 'Dog',
  breed: 'Golden Retriever',
  dateOfBirth: '2020-01-15',
  gender: 'Male',
  color: 'Golden',
  weight: 32.5,
  microchipNumber: '123456789012345',
  status: 'active',
  clientId: '1',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockPatients = [
  mockPatient,
  {
    id: '2',
    name: 'Whiskers',
    species: 'Cat',
    breed: 'Siamese',
    dateOfBirth: '2019-05-20',
    gender: 'Female',
    color: 'Cream',
    weight: 4.2,
    microchipNumber: '987654321098765',
    status: 'active',
    clientId: '1',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
  {
    id: '3',
    name: 'Max',
    species: 'Dog',
    breed: 'German Shepherd',
    dateOfBirth: '2018-03-10',
    gender: 'Male',
    color: 'Black and Tan',
    weight: 38.0,
    status: 'inactive',
    clientId: '2',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock client data
 */
export const mockClient = {
  id: '1',
  firstName: 'John',
  lastName: 'Doe',
  email: 'john.doe@example.com',
  phone: '555-0100',
  address: '123 Main St',
  city: 'Springfield',
  state: 'IL',
  zipCode: '62701',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockClients = [
  mockClient,
  {
    id: '2',
    firstName: 'Jane',
    lastName: 'Smith',
    email: 'jane.smith@example.com',
    phone: '555-0200',
    address: '456 Oak Ave',
    city: 'Springfield',
    state: 'IL',
    zipCode: '62702',
    status: 'active',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock appointment data
 */
export const mockAppointment = {
  id: '1',
  patientId: '1',
  clientId: '1',
  staffId: '1',
  appointmentType: 'checkup',
  startTime: '2025-10-25T10:00:00Z',
  endTime: '2025-10-25T10:30:00Z',
  status: 'scheduled',
  notes: 'Annual wellness exam',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockAppointments = [
  mockAppointment,
  {
    id: '2',
    patientId: '2',
    clientId: '1',
    staffId: '1',
    appointmentType: 'surgery',
    startTime: '2025-10-25T14:00:00Z',
    endTime: '2025-10-25T15:00:00Z',
    status: 'confirmed',
    notes: 'Spay procedure',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock medical record data
 */
export const mockMedicalRecord = {
  id: '1',
  patientId: '1',
  staffId: '1',
  visitDate: '2025-10-20',
  chiefComplaint: 'Annual checkup',
  diagnosis: 'Healthy',
  treatment: 'Vaccinations administered',
  prescriptions: [],
  labTests: [],
  status: 'completed',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

export const mockMedicalRecords = [
  mockMedicalRecord,
  {
    id: '2',
    patientId: '1',
    staffId: '1',
    visitDate: '2025-09-15',
    chiefComplaint: 'Ear infection',
    diagnosis: 'Otitis externa',
    treatment: 'Antibiotic ear drops',
    prescriptions: ['1'],
    labTests: [],
    status: 'completed',
    createdAt: '2025-01-01T00:00:00Z',
    updatedAt: '2025-01-01T00:00:00Z',
  },
];

/**
 * Mock prescription data
 */
export const mockPrescription = {
  id: '1',
  patientId: '1',
  medicationName: 'Amoxicillin',
  dosage: '250mg',
  frequency: 'Twice daily',
  duration: '10 days',
  instructions: 'Give with food',
  prescribedBy: '1',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock staff data
 */
export const mockStaff = {
  id: '1',
  firstName: 'Dr. Sarah',
  lastName: 'Johnson',
  email: 'sarah.johnson@purplecross.com',
  phone: '555-0300',
  role: 'veterinarian',
  specialty: 'General Practice',
  licenseNumber: 'VET-12345',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock invoice data
 */
export const mockInvoice = {
  id: '1',
  clientId: '1',
  patientId: '1',
  invoiceNumber: 'INV-2025-001',
  invoiceDate: '2025-10-20',
  dueDate: '2025-11-20',
  subtotal: 150.0,
  tax: 12.0,
  total: 162.0,
  amountPaid: 0.0,
  status: 'unpaid',
  items: [
    {
      description: 'Annual wellness exam',
      quantity: 1,
      unitPrice: 100.0,
      total: 100.0,
    },
    {
      description: 'Rabies vaccination',
      quantity: 1,
      unitPrice: 50.0,
      total: 50.0,
    },
  ],
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock inventory item data
 */
export const mockInventoryItem = {
  id: '1',
  name: 'Amoxicillin 250mg',
  category: 'Medication',
  sku: 'MED-001',
  quantity: 100,
  reorderLevel: 20,
  unitPrice: 2.5,
  supplier: 'VetSupply Co',
  expirationDate: '2026-12-31',
  status: 'active',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock lab test data
 */
export const mockLabTest = {
  id: '1',
  patientId: '1',
  testType: 'Blood panel',
  testDate: '2025-10-20',
  results: 'All values within normal range',
  status: 'completed',
  priority: 'routine',
  orderedBy: '1',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock document data
 */
export const mockDocument = {
  id: '1',
  patientId: '1',
  clientId: '1',
  title: 'Vaccination Certificate',
  documentType: 'Certificate',
  fileName: 'vaccination-cert.pdf',
  fileSize: 245678,
  mimeType: 'application/pdf',
  uploadedBy: '1',
  status: 'signed',
  createdAt: '2025-01-01T00:00:00Z',
  updatedAt: '2025-01-01T00:00:00Z',
};

/**
 * Mock user data
 */
export const mockUser = {
  id: '1',
  email: 'test@example.com',
  firstName: 'Test',
  lastName: 'User',
  role: 'veterinarian',
  token: 'mock-jwt-token',
};

/**
 * Mock API response
 */
export const mockApiResponse = <T>(data: T) => ({
  status: 'success',
  data,
});

/**
 * Mock API error response
 */
export const mockApiError = (message: string, statusCode = 400) => ({
  status: 'error',
  message,
  statusCode,
});

/**
 * Mock pagination metadata
 */
export const mockPagination = {
  page: 1,
  limit: 10,
  total: 100,
  totalPages: 10,
};

/**
 * Generate mock list with pagination
 */
export const mockPaginatedResponse = <T>(
  data: T[],
  page = 1,
  limit = 10,
  total?: number
) => ({
  status: 'success',
  data,
  pagination: {
    page,
    limit,
    total: total || data.length,
    totalPages: Math.ceil((total || data.length) / limit),
  },
});
