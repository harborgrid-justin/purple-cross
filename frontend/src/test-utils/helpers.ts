/**
 * WF-COMP-XXX | helpers.ts - Test Helper Functions
 * Purpose: Common utilities for testing
 * Dependencies: @testing-library/react
 * Last Updated: 2025-10-24 | File Type: .ts
 */

import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

/**
 * Create a user event instance for simulating user interactions
 * This should be called at the beginning of each test
 *
 * @example
 * ```tsx
 * const user = createUser();
 * await user.click(button);
 * await user.type(input, 'text');
 * ```
 */
export const createUser = () => {
  return userEvent.setup();
};

/**
 * Wait for an async operation to complete with custom timeout
 */
export const waitForAsync = async (
  callback: () => void | Promise<void>,
  timeout = 3000
): Promise<void> => {
  await waitFor(callback, { timeout });
};

/**
 * Wait for loading states to complete
 * Useful for components that show loading spinners
 */
export const waitForLoadingToFinish = async (
  getByText: (text: string | RegExp) => HTMLElement,
  timeout = 3000
): Promise<void> => {
  await waitFor(
    () => {
      expect(getByText(/loading/i)).not.toBeInTheDocument();
    },
    { timeout }
  );
};

/**
 * Mock window.scrollTo (not implemented in jsdom)
 */
export const mockScrollTo = (): void => {
  window.scrollTo = vi.fn();
};

/**
 * Mock localStorage for testing
 */
export const mockLocalStorage = (): {
  getItem: ReturnType<typeof vi.fn>;
  setItem: ReturnType<typeof vi.fn>;
  removeItem: ReturnType<typeof vi.fn>;
  clear: ReturnType<typeof vi.fn>;
} => {
  const storage: Record<string, string> = {};

  return {
    getItem: vi.fn((key: string) => storage[key] || null),
    setItem: vi.fn((key: string, value: string) => {
      storage[key] = value;
    }),
    removeItem: vi.fn((key: string) => {
      delete storage[key];
    }),
    clear: vi.fn(() => {
      Object.keys(storage).forEach((key) => delete storage[key]);
    }),
  };
};

/**
 * Mock sessionStorage for testing
 */
export const mockSessionStorage = mockLocalStorage;

/**
 * Delay execution for a specified time (in milliseconds)
 * Useful for testing loading states or debounced functions
 */
export const delay = (ms: number): Promise<void> => {
  return new Promise((resolve) => setTimeout(resolve, ms));
};

/**
 * Create a mock file for file upload testing
 */
export const createMockFile = (
  name: string,
  content: string,
  type: string
): File => {
  const blob = new Blob([content], { type });
  return new File([blob], name, { type });
};

/**
 * Create a mock image file for image upload testing
 */
export const createMockImageFile = (
  name = 'test-image.png',
  size = 1024
): File => {
  const buffer = new ArrayBuffer(size);
  const blob = new Blob([buffer], { type: 'image/png' });
  return new File([blob], name, { type: 'image/png' });
};

/**
 * Assert that an element has a specific class
 */
export const expectToHaveClass = (
  element: HTMLElement,
  className: string
): void => {
  expect(element.classList.contains(className)).toBe(true);
};

/**
 * Assert that an element does not have a specific class
 */
export const expectNotToHaveClass = (
  element: HTMLElement,
  className: string
): void => {
  expect(element.classList.contains(className)).toBe(false);
};

/**
 * Get computed style of an element
 */
export const getComputedStyleValue = (
  element: HTMLElement,
  property: string
): string => {
  return window.getComputedStyle(element).getPropertyValue(property);
};

/**
 * Simulate a network delay for testing loading states
 */
export const simulateNetworkDelay = async (ms = 500): Promise<void> => {
  await delay(ms);
};

/**
 * Create a spy for console methods (useful for testing error handling)
 */
export const spyOnConsole = (): {
  log: ReturnType<typeof vi.spyOn>;
  warn: ReturnType<typeof vi.spyOn>;
  error: ReturnType<typeof vi.spyOn>;
} => {
  return {
    log: vi.spyOn(console, 'log').mockImplementation(() => {}),
    warn: vi.spyOn(console, 'warn').mockImplementation(() => {}),
    error: vi.spyOn(console, 'error').mockImplementation(() => {}),
  };
};

/**
 * Assert that a function was called with specific arguments
 */
export const expectCalledWith = <T extends (...args: any[]) => any>(
  fn: T,
  ...args: Parameters<T>
): void => {
  expect(fn).toHaveBeenCalledWith(...args);
};

/**
 * Wait for a specific number of milliseconds in tests
 * Uses vi.advanceTimersByTime if fake timers are enabled
 */
export const advanceTime = async (ms: number): Promise<void> => {
  if (vi.isFakeTimers()) {
    vi.advanceTimersByTime(ms);
  } else {
    await delay(ms);
  }
};

/**
 * Type assertion helper for better type inference in tests
 */
export const assertType = <T>(_value: T): void => {
  // Type assertion helper - no runtime behavior
};
