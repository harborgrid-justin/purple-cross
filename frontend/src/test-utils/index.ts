/**
 * WF-COMP-XXX | index.ts - Test Utils Exports
 * Purpose: Central export point for all test utilities
 * Dependencies: None
 * Last Updated: 2025-10-24 | File Type: .ts
 */

// Export custom render functions
export * from './render';

// Export test helpers
export * from './helpers';

// Export test fixtures
export * from './fixtures';

// Export MSW mocks
export * from './mocks/handlers';
export { server, mockServer } from './mocks/server';
