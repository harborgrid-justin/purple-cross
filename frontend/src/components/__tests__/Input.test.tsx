/**
 * WF-COMP-XXX | Input.test.tsx - Input Component Tests
 * Purpose: Comprehensive tests for Input component
 * Dependencies: @testing-library/react, @testing-library/user-event
 * Last Updated: 2025-10-24 | File Type: .tsx
 */

import { describe, it, expect, vi } from 'vitest';
import { renderWithoutRouter, createUser } from '@/test-utils';
import Input from '../Input';
import { createRef } from 'react';

describe('Input Component', () => {
  describe('Rendering', () => {
    it('should render input element', () => {
      const { getByRole } = renderWithoutRouter(<Input />);

      expect(getByRole('textbox')).toBeInTheDocument();
    });

    it('should render with default classes', () => {
      const { getByRole } = renderWithoutRouter(<Input />);

      const input = getByRole('textbox');
      expect(input).toHaveClass('form-input');
    });

    it('should render form group wrapper', () => {
      const { container } = renderWithoutRouter(<Input />);

      const formGroup = container.querySelector('.form-group');
      expect(formGroup).toBeInTheDocument();
    });

    it('should render with custom className', () => {
      const { getByRole } = renderWithoutRouter(<Input className="custom-input" />);

      const input = getByRole('textbox');
      expect(input).toHaveClass('custom-input');
      expect(input).toHaveClass('form-input'); // Should still have base class
    });
  });

  describe('Label', () => {
    it('should render label when provided', () => {
      const { getByLabelText } = renderWithoutRouter(<Input label="Name" />);

      expect(getByLabelText('Name')).toBeInTheDocument();
    });

    it('should associate label with input using htmlFor', () => {
      const { getByLabelText } = renderWithoutRouter(
        <Input label="Email" id="email-input" />
      );

      const input = getByLabelText('Email');
      expect(input).toHaveAttribute('id', 'email-input');
    });

    it('should generate unique id if not provided', () => {
      const { getByLabelText } = renderWithoutRouter(<Input label="Username" />);

      const input = getByLabelText('Username');
      const inputId = input.getAttribute('id');

      expect(inputId).toBeTruthy();
      expect(inputId).toMatch(/^input-/);
    });

    it('should have correct label class', () => {
      const { container } = renderWithoutRouter(<Input label="Password" />);

      const label = container.querySelector('.form-label');
      expect(label).toBeInTheDocument();
      expect(label).toHaveTextContent('Password');
    });

    it('should not render label when not provided', () => {
      const { container } = renderWithoutRouter(<Input />);

      const label = container.querySelector('.form-label');
      expect(label).not.toBeInTheDocument();
    });

    it('should show required indicator when required', () => {
      const { container } = renderWithoutRouter(<Input label="Name" required />);

      const requiredSpan = container.querySelector('.form-required');
      expect(requiredSpan).toBeInTheDocument();
      expect(requiredSpan).toHaveTextContent('*');
    });

    it('should not show required indicator when not required', () => {
      const { container } = renderWithoutRouter(<Input label="Name" />);

      const requiredSpan = container.querySelector('.form-required');
      expect(requiredSpan).not.toBeInTheDocument();
    });
  });

  describe('Error State', () => {
    it('should render error message', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Email" error="Invalid email address" />
      );

      expect(getByRole('alert')).toHaveTextContent('Invalid email address');
    });

    it('should have error class on input', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Email" error="Invalid email" />
      );

      expect(getByRole('textbox')).toHaveClass('form-input-error');
    });

    it('should have error class with correct selector', () => {
      const { container } = renderWithoutRouter(
        <Input label="Email" error="Error" />
      );

      const errorMessage = container.querySelector('.form-error');
      expect(errorMessage).toBeInTheDocument();
      expect(errorMessage).toHaveTextContent('Error');
    });

    it('should set aria-invalid when error exists', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Email" error="Invalid" />
      );

      expect(getByRole('textbox')).toHaveAttribute('aria-invalid', 'true');
    });

    it('should set aria-invalid to false when no error', () => {
      const { getByRole } = renderWithoutRouter(<Input label="Email" />);

      expect(getByRole('textbox')).toHaveAttribute('aria-invalid', 'false');
    });

    it('should connect error message with aria-describedby', () => {
      const { getByRole, container } = renderWithoutRouter(
        <Input label="Email" id="email" error="Invalid email" />
      );

      const input = getByRole('textbox');
      const errorId = input.getAttribute('aria-describedby');
      const errorElement = container.querySelector(`#${errorId}`);

      expect(errorElement).toHaveTextContent('Invalid email');
    });

    it('should not render error when no error prop', () => {
      const { queryByRole } = renderWithoutRouter(<Input label="Email" />);

      const errors = queryByRole('alert');
      expect(errors).not.toBeInTheDocument();
    });
  });

  describe('Helper Text', () => {
    it('should render helper text', () => {
      const { getByText } = renderWithoutRouter(
        <Input label="Password" helperText="Must be at least 8 characters" />
      );

      expect(getByText('Must be at least 8 characters')).toBeInTheDocument();
    });

    it('should have helper text class', () => {
      const { container } = renderWithoutRouter(
        <Input label="Password" helperText="Helper text" />
      );

      const helper = container.querySelector('.form-helper');
      expect(helper).toBeInTheDocument();
      expect(helper).toHaveTextContent('Helper text');
    });

    it('should connect helper text with aria-describedby', () => {
      const { getByRole, container } = renderWithoutRouter(
        <Input label="Password" id="password" helperText="Helper text" />
      );

      const input = getByRole('textbox');
      const helperId = input.getAttribute('aria-describedby');
      const helperElement = container.querySelector(`#${helperId}`);

      expect(helperElement).toHaveTextContent('Helper text');
    });

    it('should not show helper text when error exists', () => {
      const { queryByText } = renderWithoutRouter(
        <Input
          label="Email"
          helperText="Enter your email"
          error="Invalid email"
        />
      );

      expect(queryByText('Enter your email')).not.toBeInTheDocument();
      expect(queryByText('Invalid email')).toBeInTheDocument();
    });

    it('should not render helper text when not provided', () => {
      const { container } = renderWithoutRouter(<Input label="Email" />);

      const helper = container.querySelector('.form-helper');
      expect(helper).not.toBeInTheDocument();
    });
  });

  describe('Full Width', () => {
    it('should apply full width class when fullWidth is true', () => {
      const { container } = renderWithoutRouter(<Input fullWidth />);

      const formGroup = container.querySelector('.form-group');
      expect(formGroup).toHaveClass('form-group-full-width');
    });

    it('should not apply full width class by default', () => {
      const { container } = renderWithoutRouter(<Input />);

      const formGroup = container.querySelector('.form-group');
      expect(formGroup).not.toHaveClass('form-group-full-width');
    });

    it('should not apply full width class when fullWidth is false', () => {
      const { container } = renderWithoutRouter(<Input fullWidth={false} />);

      const formGroup = container.querySelector('.form-group');
      expect(formGroup).not.toHaveClass('form-group-full-width');
    });
  });

  describe('Input Types', () => {
    it('should render text input by default', () => {
      const { getByRole } = renderWithoutRouter(<Input label="Name" />);

      const input = getByRole('textbox');
      expect(input.tagName).toBe('INPUT');
    });

    it('should render password input', () => {
      const { container } = renderWithoutRouter(
        <Input label="Password" type="password" />
      );

      const input = container.querySelector('input[type="password"]');
      expect(input).toBeInTheDocument();
    });

    it('should render email input', () => {
      const { container } = renderWithoutRouter(
        <Input label="Email" type="email" />
      );

      const input = container.querySelector('input[type="email"]');
      expect(input).toBeInTheDocument();
    });

    it('should render number input', () => {
      const { container } = renderWithoutRouter(
        <Input label="Age" type="number" />
      );

      const input = container.querySelector('input[type="number"]');
      expect(input).toBeInTheDocument();
    });

    it('should render tel input', () => {
      const { container } = renderWithoutRouter(
        <Input label="Phone" type="tel" />
      );

      const input = container.querySelector('input[type="tel"]');
      expect(input).toBeInTheDocument();
    });

    it('should render date input', () => {
      const { container } = renderWithoutRouter(
        <Input label="Birthday" type="date" />
      );

      const input = container.querySelector('input[type="date"]');
      expect(input).toBeInTheDocument();
    });
  });

  describe('User Interactions', () => {
    it('should accept text input', async () => {
      const user = createUser();
      const { getByRole } = renderWithoutRouter(<Input label="Name" />);

      const input = getByRole('textbox') as HTMLInputElement;
      await user.type(input, 'John Doe');

      expect(input.value).toBe('John Doe');
    });

    it('should call onChange when value changes', async () => {
      const handleChange = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Input label="Name" onChange={handleChange} />
      );

      const input = getByRole('textbox');
      await user.type(input, 'A');

      expect(handleChange).toHaveBeenCalled();
    });

    it('should call onChange for each character typed', async () => {
      const handleChange = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Input label="Name" onChange={handleChange} />
      );

      const input = getByRole('textbox');
      await user.type(input, 'Test');

      expect(handleChange).toHaveBeenCalledTimes(4); // T, e, s, t
    });

    it('should call onFocus when focused', async () => {
      const handleFocus = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Input label="Name" onFocus={handleFocus} />
      );

      const input = getByRole('textbox');
      await user.click(input);

      expect(handleFocus).toHaveBeenCalledTimes(1);
    });

    it('should call onBlur when blurred', async () => {
      const handleBlur = vi.fn();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(
        <Input label="Name" onBlur={handleBlur} />
      );

      const input = getByRole('textbox');
      await user.click(input);
      await user.tab(); // Move focus away

      expect(handleBlur).toHaveBeenCalledTimes(1);
    });

    it('should support controlled input', async () => {
      const handleChange = vi.fn();
      const user = createUser();
      const { getByRole, rerender } = renderWithoutRouter(
        <Input label="Name" value="Initial" onChange={handleChange} />
      );

      let input = getByRole('textbox') as HTMLInputElement;
      expect(input.value).toBe('Initial');

      rerender(<Input label="Name" value="Updated" onChange={handleChange} />);
      input = getByRole('textbox') as HTMLInputElement;
      expect(input.value).toBe('Updated');
    });

    it('should clear input value', async () => {
      const user = createUser();
      const { getByRole } = renderWithoutRouter(<Input label="Name" />);

      const input = getByRole('textbox') as HTMLInputElement;
      await user.type(input, 'Text to clear');
      expect(input.value).toBe('Text to clear');

      await user.clear(input);
      expect(input.value).toBe('');
    });
  });

  describe('HTML Attributes', () => {
    it('should forward placeholder attribute', () => {
      const { getByPlaceholderText } = renderWithoutRouter(
        <Input placeholder="Enter your name" />
      );

      expect(getByPlaceholderText('Enter your name')).toBeInTheDocument();
    });

    it('should forward disabled attribute', () => {
      const { getByRole } = renderWithoutRouter(<Input label="Name" disabled />);

      expect(getByRole('textbox')).toBeDisabled();
    });

    it('should forward required attribute', () => {
      const { getByRole } = renderWithoutRouter(<Input label="Name" required />);

      expect(getByRole('textbox')).toBeRequired();
    });

    it('should forward maxLength attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Name" maxLength={10} />
      );

      expect(getByRole('textbox')).toHaveAttribute('maxLength', '10');
    });

    it('should forward min and max for number inputs', () => {
      const { container } = renderWithoutRouter(
        <Input label="Age" type="number" min={0} max={120} />
      );

      const input = container.querySelector('input');
      expect(input).toHaveAttribute('min', '0');
      expect(input).toHaveAttribute('max', '120');
    });

    it('should forward pattern attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Phone" type="tel" pattern="[0-9]{3}-[0-9]{3}-[0-9]{4}" />
      );

      expect(getByRole('textbox')).toHaveAttribute(
        'pattern',
        '[0-9]{3}-[0-9]{3}-[0-9]{4}'
      );
    });

    it('should forward autoComplete attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Email" type="email" autoComplete="email" />
      );

      expect(getByRole('textbox')).toHaveAttribute('autoComplete', 'email');
    });

    it('should forward name attribute', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Username" name="username" />
      );

      expect(getByRole('textbox')).toHaveAttribute('name', 'username');
    });
  });

  describe('Ref Forwarding', () => {
    it('should forward ref to input element', () => {
      const ref = createRef<HTMLInputElement>();
      renderWithoutRouter(<Input label="Name" ref={ref} />);

      expect(ref.current).toBeInstanceOf(HTMLInputElement);
    });

    it('should allow ref to access input methods', () => {
      const ref = createRef<HTMLInputElement>();
      renderWithoutRouter(<Input label="Name" ref={ref} />);

      ref.current?.focus();
      expect(ref.current).toHaveFocus();
    });

    it('should allow ref to get input value', async () => {
      const ref = createRef<HTMLInputElement>();
      const user = createUser();
      const { getByRole } = renderWithoutRouter(<Input label="Name" ref={ref} />);

      const input = getByRole('textbox');
      await user.type(input, 'Test Value');

      expect(ref.current?.value).toBe('Test Value');
    });
  });

  describe('Accessibility', () => {
    it('should be keyboard accessible', async () => {
      const user = createUser();
      const { getByRole } = renderWithoutRouter(<Input label="Name" />);

      const input = getByRole('textbox');
      await user.tab();

      expect(input).toHaveFocus();
    });

    it('should have proper label association', () => {
      const { getByLabelText } = renderWithoutRouter(
        <Input label="Email Address" />
      );

      expect(getByLabelText('Email Address')).toBeInTheDocument();
    });

    it('should announce errors to screen readers', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Email" error="Invalid email" />
      );

      const input = getByRole('textbox');
      const errorId = input.getAttribute('aria-describedby');

      expect(errorId).toBeTruthy();
      expect(input).toHaveAttribute('aria-invalid', 'true');
    });

    it('should provide helper text to screen readers', () => {
      const { getByRole } = renderWithoutRouter(
        <Input label="Password" helperText="Must be 8+ characters" />
      );

      const input = getByRole('textbox');
      const helperId = input.getAttribute('aria-describedby');

      expect(helperId).toBeTruthy();
    });
  });

  describe('Edge Cases', () => {
    it('should handle undefined label', () => {
      const { getByRole } = renderWithoutRouter(<Input label={undefined} />);

      expect(getByRole('textbox')).toBeInTheDocument();
    });

    it('should handle empty error string', () => {
      const { getByRole } = renderWithoutRouter(<Input label="Name" error="" />);

      expect(getByRole('textbox')).toHaveAttribute('aria-invalid', 'false');
    });

    it('should handle very long error messages', () => {
      const longError =
        'This is a very long error message that should still be displayed correctly without breaking the layout';
      const { getByRole } = renderWithoutRouter(
        <Input label="Name" error={longError} />
      );

      expect(getByRole('alert')).toHaveTextContent(longError);
    });

    it('should handle special characters in value', async () => {
      const user = createUser();
      const { getByRole } = renderWithoutRouter(<Input label="Name" />);

      const input = getByRole('textbox') as HTMLInputElement;
      await user.type(input, '<script>alert("xss")</script>');

      expect(input.value).toBe('<script>alert("xss")</script>');
    });
  });

  describe('Snapshot Tests', () => {
    it('should match snapshot for basic input', () => {
      const { container } = renderWithoutRouter(<Input label="Name" />);
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for input with error', () => {
      const { container } = renderWithoutRouter(
        <Input label="Email" error="Invalid email" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });

    it('should match snapshot for required input with helper text', () => {
      const { container } = renderWithoutRouter(
        <Input label="Password" required helperText="Must be 8+ characters" />
      );
      expect(container.firstChild).toMatchSnapshot();
    });
  });
});
