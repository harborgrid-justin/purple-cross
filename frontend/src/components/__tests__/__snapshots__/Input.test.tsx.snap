// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Input Component > Snapshot Tests > should match snapshot for basic input 1`] = `
<div
  class="form-group"
>
  <label
    class="form-label"
    for="input-jznsq"
  >
    Name
  </label>
  <input
    aria-invalid="false"
    class="form-input"
    id="input-jznsq"
  />
</div>
`;

exports[`Input Component > Snapshot Tests > should match snapshot for input with error 1`] = `
<div
  class="form-group"
>
  <label
    class="form-label"
    for="input-i24uw"
  >
    Email
  </label>
  <input
    aria-describedby="input-i24uw-error"
    aria-invalid="true"
    class="form-input form-input-error"
    id="input-i24uw"
  />
  <p
    class="form-error"
    id="input-i24uw-error"
    role="alert"
  >
    Invalid email
  </p>
</div>
`;

exports[`Input Component > Snapshot Tests > should match snapshot for required input with helper text 1`] = `
<div
  class="form-group"
>
  <label
    class="form-label"
    for="input-kxb6jj"
  >
    Password
    <span
      class="form-required"
    >
      *
    </span>
  </label>
  <input
    aria-describedby="input-kxb6jj-helper"
    aria-invalid="false"
    class="form-input"
    id="input-kxb6jj"
    required=""
  />
  <p
    class="form-helper"
    id="input-kxb6jj-helper"
  >
    Must be 8+ characters
  </p>
</div>
`;
